rekcurd/data_servers/__init__.py

```diff
@@ -115,19 +115,26 @@ def upload_evaluation_data(self, request_iterator: Iterator[rekcurd_pb2.UploadEv
         self._api_handler.upload(filepath, str(local_filepath))
         return str(local_filepath)
 
-    def _upload_evaluation_result(self, data: object, filepath: str, suffix: str):
+    def upload_evaluation_result(self, data_gen: Generator[EvaluateResultDetail, None, EvaluateResult], filepath: str) -> EvaluateResult:
         valid_path = convert_to_valid_path(filepath)
         if filepath != str(valid_path):
             raise Exception(f'Error: Invalid evaluation result file path specified -> {filepath}')
-        local_filepath = Path(self._api_handler.LOCAL_EVAL_DIR, valid_path.name+suffix)
-        local_filepath.parent.mkdir(parents=True, exist_ok=True)
-        with local_filepath.open(mode='wb') as f:
-            pickle.dump(data, f)
-        self._api_handler.upload(filepath+suffix, str(local_filepath))
-        return str(local_filepath)
 
-    def upload_evaluation_result_summary(self, data: object, filepath: str) -> str:
-        return self._upload_evaluation_result(data, filepath, self.__EVALUATE_RESULT)
+        local_dir = Path(self._api_handler.LOCAL_EVAL_DIR)
+        local_dir.mkdir(parents=True, exist_ok=True)
+
+        detail_local_filepath = local_dir / (valid_path.name+self.__EVALUATE_DETAIL)
+        with detail_local_filepath.open(mode='wb') as detail_file:
+            try:
+                while True:
+                    pickle.dump(next(data_gen), detail_file)
+            except StopIteration as e:
+                result_local_filepath = local_dir / (valid_path.name+self.__EVALUATE_RESULT)
+                evaluate_result = e.value
```

Comment:
I am not familiar with `StopIteration`.
Does this mean that when `StopIteration` exception happens, will `e` be `EvaluateResult` object? Sounds interesting.

Reply:
Yes, (actually `e.value`, not `e` )
This is not a normal way, but I chose this to make implementation of `evaluate` method simple.

Usually, `return` statement in generator is used 

- to occur StopIteration explicitly
- to tell error message.

e.g.

```
def gen() -> Generator[int, None, str]:
    yield 10
    yield 20
    return 'no more values'

try:
    g = gen()
    while True:
        print(next(g))
except StopIteration as e:
    print(e) # <- print  'no more values'
```

Reply:
Thank you for your clarification.
And you implemented your `evaluate` example like [this way](https://github.com/rekcurd/rekcurd-example/blob/61ece0cfb7fc0cbb97c4845c9823e69d28b03628/python/sklearn-digits/app.py#L65-L99). It is a great job.